src/io/mod.rs

```diff
@@ -117,9 +117,14 @@ pub trait Write {
     fn write_all(&mut self, mut buf: &[u8]) -> Result<()> {
         while !buf.is_empty() {
             match self.write(buf) {
-                Ok(0) => return Err(Error::new(ErrorKind::WriteZero,
-                                               "failed to write whole buffer")),
-                Ok(n) => buf = &buf[n..],
+                Ok(0) => return Err(Error::from(ErrorKind::WriteZero)),
```

Comment:
Could you add a NOTE(steed) here to indicate that we are diverging from upstream to remove an allocation at the cost a custom error message?